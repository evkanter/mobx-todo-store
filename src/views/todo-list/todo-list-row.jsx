import React from 'react';
import { inject, observer } from 'mobx-react';
import PropTypes from 'prop-types';

@inject('store')
@observer 
export class TodoListRow extends React.Component {

	static propTypes = {
        num: PropTypes.number.isRequired,
        todoName: PropTypes.string.isRequired,
        enteredDate: PropTypes.string.isRequired,
        store: PropTypes.any,
        isCompleted: PropTypes.bool.isRequired
	}

	constructor(props) {
		super(props);
		this.store = this.props.store.todoListStore;
        this.onToggleTodoCheckbox = this.onToggleTodoCheckbox.bind(this);
	}

    onToggleTodoCheckbox(index) {
		this.store.toggleIsCompletedValueByIndex(index);
    }

	render () {
		return(
            <tr>
                <td>{ this.props.enteredDate }</td>
                <td>{ this.props.todoName }</td>
                <td className="checkbox-container"><input type="checkbox" 
                    name={'todo-'+this.props.num} 
                    onChange={this.onToggleTodoCheckbox.bind(this,this.props.num)} 
                    checked={this.props.isCompleted}
                /></td>
            </tr>
		);
	}
}