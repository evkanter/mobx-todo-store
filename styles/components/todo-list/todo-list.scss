.todo-list-container {
    width: $fn-container-width;
    height: auto;
    margin: 0 auto;
    padding: 10px;
    position: relative;
}
.todo-list {
    background-color: $fn-todo-table-background-color;
    border: 1px solid $fn-todo-table-color;
    color: $fn-todo-table-color;
    text-align: center;
    border-radius: 10px 10px 0 0;
    h1 {
        font-size: 16px;
        font-weight: bold;
    }
    table {
        width: 100%;
        background-color: $fn-todo-table-background-color;
        border-top: 1px solid $fn-todo-table-color;
        color: $fn-todo-table-color;
        text-align: center;
        border-collapse:separate; 
        border-spacing:0 0;
        thead {
            background-color: $fn-todo-table-header-color;
            color: $fn-black;
            text-align: center;
        }
        tbody {
            background-color: $fn-todo-table-background-color;
            color: $fn-black;
            text-align: center;
            tr > td:not(:last-child) {
                border-right: 1px solid $fn-todo-table-color;
            }
            tr > td {
                border-top: 1px solid $fn-todo-table-color;
            }
            tr.grey-row > td {
                background-color: #999999;
                border-right: 0;
            }
        }
    }
    .small {
        width: $fn-row-width-small;
    }
    .button-bar {
        text-align: right;
    }
    button {
        width: 200px;
        border: 1px solid #000000;
        margin: 5px;
        padding: 5px;
        text-align:center;
    }
    button.small {
        width: $fn-row-width-button-small;
        padding: 8px;
    }
    button:disabled {
        border: 1px solid #666666;
    }
    input[type='text'], input[type='date'] {
        border:1px solid #000000;
        padding:5px;
        display: inline-block;
        float: left;
        height: 20px;
        margin:5px;
        background-color: #FFFFFF;
        text-align: center;
        &.fullsize {
            width:94%;
        }
      }
    input[type='checkbox'] {
        border:1px solid #000000;
        width: 20px;
        background-color: #FFFFFF;
        text-align: center;
        margin:5px;    
    }
    td.checkbox-container {
        text-align: center;
        vertical-align: middle;
        border-left: 0;
    }
}