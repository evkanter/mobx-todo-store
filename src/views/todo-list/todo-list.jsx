import React from 'react';
import { observer, inject, PropTypes } from 'mobx-react';

import { TodoListRow } from './todo-list-row.jsx';

@inject('store')
@observer 
export class TodoList extends React.Component {

	static propTypes = {
		store: PropTypes.observableObject.isRequired
	}

	constructor(props) {
		super(props);
		this.store = this.props.store.todoListStore;
		this.state = {newTodoName: '', newTodoDate: ''};

		this.addATodo = this.addATodo.bind(this);
		this.clearCompletedTodos = this.clearCompletedTodos.bind(this);
		this.handleTextboxChange = this.handleTextboxChange.bind(this);
		this.resetInputFields = this.resetInputFields.bind(this);
	}

	addATodo(event) {
		event.preventDefault();
		let todo = {
			name: this.state.newTodoName, 
			date: this.state.newTodoDate
		};
		this.store.addNewTodoToList(todo);
		this.resetInputFields();
	}

	clearCompletedTodos(event) {
		event.preventDefault();
		this.store.clearCompletedTodos();
	}
	
	handleTextboxChange (name, e) {
		if (name==='newTodoName') {
			this.setState({newTodoName: e.target.value});
		} else if (name==='newTodoDate') {
			this.setState({newTodoDate: e.target.value});
		}
	}

	resetInputFields() {
		this.setState({newTodoName: '', newTodoDate: ''});
	}
	
	getTodos() {
		this.rows = this.store.todoList.map(function(todo, i){
			return <TodoListRow 
				key={i} 
				num={i} 
				enteredDate={todo.enteredDate} 
				todoName={todo.todoName} 
				isCompleted={todo.isCompleted} />;
		}, this.props)
	}

	render () {
		this.getTodos();
		return(
			<section className="todo-list-container">
				<div className="todo-list">
					<h1>My Todo List</h1>
					<form>
						<table>
							<thead>
								<tr>
									<th className="small">date</th>
									<th>to do</th>
									<th className="small">&#10003;</th>
								</tr>
							</thead>
							<tbody>

								{this.rows}

								<tr className="grey-row">
									<td><input type="date" 
											id="newTodoDate" 
											value={this.state.newTodoDate}
											onChange={this.handleTextboxChange.bind(this,'newTodoDate')}																		
									/></td>
									<td><input type="text" 
											id="newTodoName" 
											className="fullsize"
											placeholder="Add an item"
											value={this.state.newTodoName} 
											maxLength="60"
											required="required"
											onChange={this.handleTextboxChange.bind(this,'newTodoName')}								
									/></td>
									<td><button className="small"
										onClick={this.addATodo.bind(this)} 
										disabled={this.state.newTodoName.length === 0} >+</button></td>
								</tr>

								<tr>
									<td colSpan={3} className="button-bar">
										<button onClick={this.clearCompletedTodos.bind(this)}>Clear completed</button>
									</td>
								</tr>
							</tbody>
						</table>
						
					</form>
				</div>
            </section>
		);
	}
}