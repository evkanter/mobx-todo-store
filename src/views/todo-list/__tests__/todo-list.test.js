jest.unmock('../todo-list');

import { TodoList } from '../todo-list';
import { MemoryRouter } from 'react-router-dom';
import { todoListStore } from '../../../core/stores/todo-list.store.js'

describe('<TodoList />', () => {

    const store = todoListStore;

    describe('renders', () => {
        test('matches previous todo list snapshot', () =>{
            const component = renderer.create(
                <MemoryRouter>
                    <TodoList store={store} />
                </MemoryRouter>
            );
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});
