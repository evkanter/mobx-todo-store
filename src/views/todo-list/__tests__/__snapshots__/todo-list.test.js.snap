// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> renders matches previous todo list snapshot 1`] = `
<section
  className="todo-list-container"
>
  <div
    className="todo-list"
  >
    <h1>
      My Todo List
    </h1>
    <form>
      <table>
        <thead>
          <tr>
            <th
              className="small"
            >
              date
            </th>
            <th>
              to do
            </th>
            <th
              className="small"
            >
              âœ“
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td />
            <td />
            <td>
              <input
                name="todo-1"
                type="checkbox"
                value="checked"
              />
            </td>
          </tr>
          <tr>
            <td />
            <td />
            <td>
              <input
                name="todo-2"
                type="checkbox"
                value="checked"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </form>
  </div>
</section>
`;
